GoogleForCreators/web-stories-wp
# Border radius panel

## Summary
Adds border radius panel


## Relevant Technical Choices
Border-radius together with border came out to be more complicated than expected, especially for the outside border.
Outside border can't be a fully separate element as in case of inner border for the following reason:
- If the outside border is a separate element and we are assigning border-radius for both the outside border element and the "real" element, the border radiuses will not match and there will be room between the border and the element itself.
- We can't apply the border and radius directly to the inner element either + adjust the size of it (make it larger to cover the border area and then apply border) since this would not work with zoom. Also, the mask logic would clip anything overflowing the mask area anyway.

What this PR currently does to solve this is  
- Increase the size of the maks element so that it covers the additional border area
- Applies border-radius and border to the wrapper of the inner element to still allow zooming

This, however, leads to an additional problem with resizing -- since the resizing does not change the style of the wrapper element currently, then the outer border is not adjusting to resizing -- this still needs to be figured out.



## To-do

- [x] Outside border: Support for text highlight mode
- [x] Outside border: support for resizing
- [x] Outside border: support for video
- [x] Outside border: support for shape
-  Moveable radius handle -- NOTE: will be split out to a separate issue.
- [x] Support for shape element


## User-facing changes



## Testing Instructions
I
- Add an element (test with all element types)
- Unlock the border radius and assign values, verify it gets assigned as expected to all corners
- Lock the radius, verify that all the values are unified to match the first border radius input
- Change one value, verify that all the other values get changed, too
- Enter edit mode (if applicable), verify that the radius is visible there, too.

II
- Test border-radius together with border for different element types.
- Try unlocking the border and border-radius, assign different values to each corner/side, and test these together.
- It's good to test with border that has opacity -- this way the center border can be better seen.
- Test through all types: inside, center and outside border with border radius.
- Verify that it all works together.
- Verify that resizing elements with border + border-radius works well together.

III 
Verify that all this looks well in the output (preview / front-end) as well!

Note that there seems to be an unrelated issue with the text element output: https://github.com/google/web-stories-wp/issues/5324 that influences how the border looks like on the text element.



---



Fixes #4850
